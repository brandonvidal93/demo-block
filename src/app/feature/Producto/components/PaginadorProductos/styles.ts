import styled from 'styled-components';

export const NavContainer = styled.nav`
  display: flex;
  align-items: center;
  justify-content: space-evenly;
`;

export const ButtonNav = styled.button`
  background: #B1E6C8;
  border: 0px;
  border-radius: 3px;
  width: 20px;
  padding: 0.1rem;
  cursor: pointer;
  transition: all 0.2s ease-out;

  &:hover {
    background: #69E09D;
  }
`;