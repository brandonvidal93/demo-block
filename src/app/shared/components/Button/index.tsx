import styled from 'styled-components';

export const Button = styled.button`
  background: #51AD79;
  border-radius: 3px;
  border: 0px;
  color: #fff;
  height: 32px;
  display: block;
  width: 100%;
  text-align: center;
  cursor: pointer;
  transition: all 0.2s ease-out;

  &[disabled] {
    opacity: 0.3;
  }
  
  &:hover {
    background: #2E6144;
  }
`;

export const ButtonDanger = styled.button`
  background: #dc3545;
  border-radius: 3px;
  border: 0px;
  color: #fff;
  height: 32px;
  display: block;
  width: 100%;
  text-align: center;
  cursor: pointer;
  transition: all 0.2s ease-out;

  &[disabled] {
    opacity: 0.3;
  }
  
  &:hover {
    background: #bb2d3b;
  }
`;
