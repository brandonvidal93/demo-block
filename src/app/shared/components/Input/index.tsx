import styled from 'styled-components';

export const Input = styled.input`
  border: 1px solid #51AD79;
  border-radius: 3px;
  margin-bottom: 1px;
  padding: 8px 4px;
  display: block;
  width: 100%;
  &[disabled] {
    opacity: 0.3;
  }
`;
