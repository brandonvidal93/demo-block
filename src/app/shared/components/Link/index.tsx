import { Link as RouterLink } from 'react-router-dom';
import styled from 'styled-components';

export const Link = styled(RouterLink)`
  text-decoration: none;
  color: #4B6154;
  &:hover {
    color: #2E6144;
  }
`;
