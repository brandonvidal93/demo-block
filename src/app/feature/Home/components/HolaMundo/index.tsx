import * as PropTypes from 'prop-types';
import * as React from 'react';
import { DivContainer, DivSection, H1Title, H2Title, H3Title, LiText, TextBody, UlList } from './styles';


interface HolaMundoProps {
  msg: string;
}

export const HolaMundo: React.FC<HolaMundoProps> = ({ msg }) => 
{
  return(
    <DivContainer>
      <div>
        <H1Title>{msg}</H1Title>
      </div>

      <div>
        <H2Title>Estructura de Código</H2Title>
        <TextBody>Organizada, mantenible y escalable</TextBody>
      </div>

      <DivSection>
        <H3Title>Directorios</H3Title>

        <TextBody>
          <b>Core: </b>Núcleo de la aplicación o configuración principal, la cual controla o estable el comportamiento de la aplicación.
        </TextBody>

        <TextBody>
          <b>Feature:</b> Módulo principales del proyecto, cada uno de ellos contiene la funcionalidad específica y que agrega valor a la totalidad del proyecto.
        </TextBody>

        <TextBody>
          <b>Shared:</b> Componentes, utilitarios y/o recursos que se pueden utilizar en toda la aplicación, es decir, son compartidos por los módulos principales.
        </TextBody>
      </DivSection>

      <DivSection>
        <H3Title>Buenas prácticas de desarrollo</H3Title>

        <UlList>
          <LiText>
            <a href='https://dev.to/abrahamlawson/react-style-guide-24pp'>Manual de estilos</a>
          </LiText>
          <LiText>Versionamiento del código</LiText>
          <LiText>Código limpio</LiText>
          <LiText>Pruebas</LiText>
        </UlList>
      </DivSection>

      <DivSection>
        <H2Title>Proceso</H2Title>
        <TextBody>
          El proceso de desarrollo se lleva a cabo en varias etapas, las cuales nos garantizan que el código será desplegado en los ambientes con la mejor calidad y asegurando un correcto funcionamiento.
        </TextBody>

        <UlList>
          <LiText>Programación</LiText>
          <LiText>Pruebas</LiText>
          <LiText>Integración (CI)</LiText>
          <LiText>Despliegue (CD)</LiText>
        </UlList>
      </DivSection>
    </DivContainer>
  );
};

HolaMundo.propTypes = {
  msg: PropTypes.string.isRequired,
};
