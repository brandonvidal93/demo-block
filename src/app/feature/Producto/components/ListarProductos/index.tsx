import * as PropTypes from 'prop-types';
import * as React from 'react';
import { Table, TextTable, TitleTableBold } from './styles';
import { BtnEliminarProducto } from '../EliminarProducto';
import { Producto } from '../../models/Producto';

export interface ListaProductosProps {
  productos: Array<Producto>;
  onClickEliminarProducto: (productos: Producto) => void;
}

export const ListaProductos: React.FC<ListaProductosProps> = ({
  productos,
  onClickEliminarProducto,
}) => {
  return (
    <Table>
      <thead>
        <tr>
          <td>
            <TitleTableBold>Título</TitleTableBold>
          </td>
          <td>
            <TitleTableBold>Fecha creación</TitleTableBold>
          </td>
          <td>
            <TitleTableBold>Eliminar</TitleTableBold>
          </td>
        </tr>
      </thead>
      <tbody>
        {productos.map((producto: Producto) => {
          return (
            <tr key={producto.slug}>
              <TextTable>{producto.title}</TextTable>
              <TextTable>{`${producto.createdAt} `}</TextTable>
              <td>
                <BtnEliminarProducto
                  producto={producto}
                  onEliminar={onClickEliminarProducto}
                ></BtnEliminarProducto>
              </td>
            </tr>
          );
        })}
      </tbody>
    </Table>
  );
};

ListaProductos.propTypes = {
  productos: PropTypes.array.isRequired,
  onClickEliminarProducto: PropTypes.func.isRequired,
};
