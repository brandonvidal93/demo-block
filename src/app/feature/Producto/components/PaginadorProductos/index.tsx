import * as PropTypes from 'prop-types';
import * as React from 'react';
import { ButtonNav, NavContainer } from './styles';

interface PaginadorProductosProps {
  cantidadTotalProductos: number;
  onClickCambiarPagina: (numeroPagina: number) => void;
}

const PRODUCTOS_VISIBLES_POR_PAGINA = 10;

export const PaginadorProductos: React.FC<PaginadorProductosProps> = ({
  onClickCambiarPagina,
  cantidadTotalProductos,
}) => {
  if (cantidadTotalProductos <= PRODUCTOS_VISIBLES_POR_PAGINA) {
    return null;
  }

  const rango = Array.from(
    Array(
      Math.ceil(cantidadTotalProductos / PRODUCTOS_VISIBLES_POR_PAGINA)
    ).keys()
  );

  return (
    <NavContainer>
      {rango.map((index) => {
        return (
          <ButtonNav
            onClick={() => onClickCambiarPagina(index)}
            key={index.toString()}
          >
            {index + 1}
          </ButtonNav>
        );
      })}
    </NavContainer>
  );
};

PaginadorProductos.propTypes = {
  cantidadTotalProductos: PropTypes.number.isRequired,
  onClickCambiarPagina: PropTypes.func.isRequired,
};
