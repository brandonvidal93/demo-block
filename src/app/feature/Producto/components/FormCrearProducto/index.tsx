import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as Yup from 'yup';
import { DivContainer, DivContainerButton, SpanError, TitleH2 } from './styles';
import { Button } from 'app/shared/components/Button';
import { FormikHelpers } from 'formik/dist/types';
import { Input } from 'app/shared/components/Input';
import { Producto } from '../../models/Producto';
import { useFormik } from 'formik';

interface FormValues {
  title: string;
  slug: string;
  body: string;
}

interface FormCrearProductoProp {
  onSubmit: (payload: Producto) => any;
  disabled?: boolean;
  formTitle: string;
  initialValues?: FormValues;
}

const validationSchema = Yup.object().shape<FormValues>({
  title: Yup.string().required('El campo title es requerido.'),
  slug: Yup.string().required('El campo slug es requerido.'),
  body: Yup.string().required('El campo body es requerido.'),
});

export const FormCrearProducto: React.FC<FormCrearProductoProp> = ({
  onSubmit,
  disabled,
  formTitle,
  initialValues = {
    title: '',
    slug: '',
    body: '',
  },
}) => {
  const handleSubmit = (
    values: FormValues,
    { resetForm }: FormikHelpers<FormValues>
  ) => {
    onSubmit({
      title: values.title,
      slug: values.slug,
      body: values.body,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    });
    resetForm();
  };
  const formik = useFormik({
    initialValues,
    validationSchema,
    onSubmit: handleSubmit,
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <TitleH2>{formTitle}</TitleH2>
      <DivContainer>
        <Input
          disabled={disabled}
          name="title"
          placeholder="Title"
          value={formik.values.title}
          onChange={formik.handleChange}
        />
        {formik.touched.title && formik.errors.title && (
          <SpanError>{formik.errors.title}</SpanError>
        )}
      </DivContainer>
      <DivContainer>
        <Input
          disabled={disabled}
          name="slug"
          placeholder="Slug"
          value={formik.values.slug}
          onChange={formik.handleChange}
        />
        {formik.touched.slug && formik.errors.slug && (
          <SpanError>{formik.errors.slug}</SpanError>
        )}
      </DivContainer>
      <DivContainer>
        <Input
          disabled={disabled}
          name="body"
          placeholder="Body"
          value={formik.values.body}
          onChange={formik.handleChange}
        />
        {formik.touched.body && formik.errors.body && (
          <SpanError>{formik.errors.body}</SpanError>
        )}
      </DivContainer>
      <DivContainerButton>
        <Button type="submit">Registrar</Button>
      </DivContainerButton>
    </form>
  );
};

FormCrearProducto.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  formTitle: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  initialValues: PropTypes.shape({
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired,
    slug: PropTypes.string.isRequired,
  }),
};
