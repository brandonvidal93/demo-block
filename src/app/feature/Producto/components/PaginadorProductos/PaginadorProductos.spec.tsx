import * as React from 'react';
import {render, screen} from '@testing-library/react';
import { PaginadorProductos } from 'app/feature/Producto/components/PaginadorProductos';

describe('PaginadorProductos Test', () => {


  it('should match snapshot', () => {
    const { container }  = render(
      <PaginadorProductos
        cantidadTotalProductos={11}
        onClickCambiarPagina={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('Renderizar con menos de 10 productos no debe pintar botones', () => {
    const { container } = render(
      <PaginadorProductos
        cantidadTotalProductos={9}
        onClickCambiarPagina={() => {}}
      />
    );

    expect(container).toBeTruthy();
  });

  it('Renderizar con  20 productos debe pintar 2 botones', () => {
    const { container } = render(
      <PaginadorProductos
        cantidadTotalProductos={20}
        onClickCambiarPagina={() => {}}
      />
    );
    const buttons = screen.getAllByRole('button');
    expect(buttons[0]).toBeTruthy();
    expect(buttons[1]).toBeTruthy();
    
  });
});

export {};
