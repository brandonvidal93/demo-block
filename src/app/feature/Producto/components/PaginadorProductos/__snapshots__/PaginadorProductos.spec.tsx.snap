// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginadorProductos Test should match snapshot 1`] = `
<div>
  <nav>
    <button>
      1
    </button>
    <button>
      2
    </button>
  </nav>
</div>
`;
